From 66d5cdaa8020e10b3fa0e16ca3233a3b2210c092 Mon Sep 17 00:00:00 2001
From: "jinrong.liao" <jinrong.liao@amlogic.com>
Date: Fri, 15 Jan 2021 13:26:37 +0800
Subject: [PATCH 1/5] env: add env support for nand/emmc for uboot 2019  [1/1]

PD#SWPL-41293

Problem:
Never use mtd part as env in amlogic case;

Solution:
Never use mtd part as env in amlogic case;
For Nand/Emmc base, we use /dev/nand_env and /dev/block/env.

Verify:
verify by A1

Signed-off-by: jinrong.liao <jinrong.liao@amlogic.com>
Change-Id: I4693ffb5a0a813547379036779277978bfc4a919
---
 tools/env/fw_env.c      | 18 ++++++++++++++++--
 tools/env/fw_env.config |  6 ++++++
 2 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/tools/env/fw_env.c b/tools/env/fw_env.c
index a5d75958e1..406a7cd660 100644
--- a/tools/env/fw_env.c
+++ b/tools/env/fw_env.c
@@ -1181,6 +1181,9 @@ static int flash_write(int fd_current, int fd_target, int dev_target)
 		DEVOFFSET(dev_target), DEVNAME(dev_target));
 #endif

+	/* Amlogic Add */
+	DEVTYPE(dev_target) = MTD_ABSENT;
+
 	if (IS_UBI(dev_target)) {
 		if (ubi_update_start(fd_target, CUR_ENVSIZE) < 0)
 			return 0;
@@ -1217,6 +1220,17 @@ static int flash_read(int fd)
 		return ubi_read(fd, environment.image, CUR_ENVSIZE);
 	}

+	/*
+	 * Never use mtd part as env in amlogic case;
+	 * For Nand/Emmc base, we use chardev '/dev/nand_env',
+	 * and /dev/block/env.
+	 */
+	struct mtd_info_user mtdinfo;
+	memset(&mtdinfo, 0, sizeof(mtdinfo));
+	mtdinfo.type = MTD_ABSENT;
+
+	DEVTYPE(dev_current) = mtdinfo.type;
+
 	rc = flash_read_buf(dev_current, fd, environment.image, CUR_ENVSIZE,
 			    DEVOFFSET(dev_current));
 	if (rc != CUR_ENVSIZE)
@@ -1621,14 +1635,14 @@ static int check_device_config(int dev)
 				DEVNAME(dev));
 			goto err;
 		}
-		if (mtdinfo.type != MTD_NORFLASH &&
+/*		if (mtdinfo.type != MTD_NORFLASH &&
 		    mtdinfo.type != MTD_NANDFLASH &&
 		    mtdinfo.type != MTD_DATAFLASH &&
 		    mtdinfo.type != MTD_UBIVOLUME) {
 			fprintf(stderr, "Unsupported flash type %u on %s\n",
 				mtdinfo.type, DEVNAME(dev));
 			goto err;
-		}
+		}*/
 		DEVTYPE(dev) = mtdinfo.type;
 		if (DEVESIZE(dev) == 0)
 			/* Assume the erase size is the same as the env-size */
diff --git a/tools/env/fw_env.config b/tools/env/fw_env.config
index 053895a2c0..65a4acf344 100644
--- a/tools/env/fw_env.config
+++ b/tools/env/fw_env.config
@@ -36,3 +36,9 @@
 # UBI volume by name
 #/dev/ubi0:env		0x0		0x1f000		0x1f000
 #/dev/ubi0:env-redund	0x0		0x1f000		0x1f000
+
+# Amlogic nand example, the env size need to be set according to board config
+/dev/nand_env		0x0		0x10000		0x200
+
+# Amlogic emmc example
+/dev/block/env		0x0		0x40000		0x200
--
2.27.0

